### ISSUE TITLE
Automated upgrades... not so automated...

### ISSUE THREAD

**Issue by Pontiac76** [kasmtech/workspaces-issues#290]
I've got Kasm setup on my own domain on my home servers.  It runs for four users.  Me... Myself.. and I... and the admin user just to maintain the software.

Today, I attempted to kick up Discord through Kasm for the first time using my non-admin account, but I ran into a loop with it telling me that there was an upgrade, it gave me three options, but, I couldn't get any further than that.  I figure that it's just something with permissions of the user I'm logging in with, so, went on the hunt to figure out how to update EVERYTHING in Kasm since it's been a couple months from initial install.

I ran across the Single Server Upgrade page https://www.kasmweb.com/docs/latest/upgrade/single_server_upgrade.html
If I go to https://kasm-static-content.s3.amazonaws.com/ I get an access denied (I fully expected that) so I can't get a list of files that are available to grab from that S3 bucket.  The random 8-byte at the end of the filename also makes it kind of difficult to do any sort of pull in the future unless I monitor that exact HTML page.

If there's something available that'll GET me the most current version from a command line automatically, please let me know and we can close this.  If it's just something here on Git or if it's listed in a text file I can pull, I'm all ears.

But if not, I'd like to offer a couple of suggestions to bring this to a point that it is automated...

- Point me to something in Github that provides me with whatever the recent release is.  I can work with something from there either from a scrape or from some other API or whatever.
- Have a text file (CSV/TSV) available that contains all of the files available to download from the bucket with the age of the file.  I can then pull that CSV (lets call it) and then figure out what I want to do from there.
- Keep the different revisions of the files that exist now, but, also copy the most recent version to a static filename that'll always exist.

I appreciate this tool.  It's come in useful a couple of times already when I'm at work as well as when I'm at my parents place.

**Comment by j-travis** [comment]
If your immediate concern is about discord being out of date , you might consider updating your image to point to the "rolling" edition which is built is published nightly

For example, workspaces 1.11 would install the discord image as `kasmweb/discord:1.11.0` .
If you change that to `kasmweb/discord:1.11.0-rolling` your system would auto pull a new version of discord every day. The same goes for all our other images.

Ref: 
- https://kasmweb.com/docs/latest/guide/custom_images.html#rolling-docker-images
- https://hub.docker.com/r/kasmweb/discord

**Comment by Pontiac76** [comment]
I had the itch to figure this out... But I don't love the idea of relying on the download page for stuff like this as pages like that change regularly.

After installing lynx on my Debian machine, I came up with this:

lynx -dump -listonly https://www.kasmweb.com/downloads | grep "release_[[:digit:]]\." | cut -f2- -d"."
 https://kasm-static-content.s3.amazonaws.com/kasm_release_1.12.0.d4fd8a.tar.gz
 https://kasm-static-content.s3.amazonaws.com/kasm_release_1.11.0.18142e.tar.gz
 https://kasm-static-content.s3.amazonaws.com/kasm_release_1.10.0.238225.tar.gz
 https://kasm-static-content.s3.amazonaws.com/kasm_release_1.9.0.077388.tar.gz

It's a top to bottom list of what to grab, so that's "nice" but that's not a guarantee that'll be the same in the future.

@j-travis 
Thanks.  I'll look at those links.

But before I do, wouldn't this still runs into a version problem later on down the road?  What does 1.11 represent exactly?  Is that reference to Kasm 1.11?  Is that something specific to Discord?  I'm assuming Kasm, which would mean I'd need to change what you put here to 1.12 if I were to do the full release, but then there's the reliance on trying to keep tabs on what the current version of Kasm is installed.  If there's something on the OS FS that'll tell me what the current release is of Kasm, I'm more comfortable with the rolling idea.

Is 1.11 the current CE version while 1.12 is the commercial version?

.. and.. if 1.11 is the CE version and 1.12 is the commercial, what's going to happen if I do the full upgrade from 1.11 to 1.12 if I'm on the CE version?

**Comment by j-travis** [comment]
1.11 represents the release version of Kasm Workspaces platform. We version the  [workspace images](https://github.com/kasmtech/workspaces-images) -e.g chrome / discord according to the version of Workspaces they built and compatible with.

Thats why you will see tags for `1.10.0, 1.11.0, 1.12.0` and then their respective "rolling" editions.  The target app  (e.g discord) inside the rolling images is updated but the wiring inside the image that makes it compatible with workspaces remains unchanged so that you can ensure compatibility.

This is precisely why you dont see us creating `latest` tags for those, because sometimes there are breaking changes between releases of Kasm Workspaces.

We just published 1.12 within the last few days so its currently the newest release of CE and commercial version . Its all the same build - just some features require a license.

Upgrades from one version to another are documented.

Sorry, aside from the download page on our site, there is no special URL that reports the latest version and its download link. 

Workspaces is sufficiently complex, that I caution against writing automation to auto fetch and install/upgrade new versions of the platform.  Install / upgrade procedures change from version to version so you're just asking for things to break.

**Comment by Pontiac76** [comment]
I decided instead of just upgrading the workspace images, I'd upgrade the entire package.  The install SEEMS to have gone right, no errors other than some missing packages apparently during the cleanup process, but now I can't log in as either my normal user I created and the admin account.

Checking the logs, I see that it did some big time deletes, like, dropping the database.  ... ok?

I can see that the installer created a new set of creds for admin@kasm.local.  But I don't seem to be able to log in as that.

Going down to the near the bottom of the log file on my console, I see this:

```
2022-12-06 17:07:20,263 [INFO] root: Performing Database Connectivity Test
Traceback (most recent call last):
  File "server.py", line 218, in <module>
IsADirectoryError: [Errno 21] Is a directory: '/tmp/default_images_unknown.yaml'
[1] Failed to execute script 'server' due to unhandled exception!

```

Looking for "default"

```root@Kasm:/tmp# find /tmp -type f | grep default
/tmp/kasm_release/conf/database/seed_data/default_agents.yaml
/tmp/kasm_release/conf/database/seed_data/default_connection_proxies.yaml
/tmp/kasm_release/conf/database/seed_data/default_images_amd64.yaml
/tmp/kasm_release/conf/database/seed_data/default_properties.yaml
/tmp/kasm_release/conf/database/seed_data/default_images_arm64.yaml```

Seems like something missed its target?

**Comment by Pontiac76** [comment]
After the reboot, and then noticing that the installer moves the default port to somewhere else, my proxy is happy, and I can log in.  So it looks like there's some kind of restore after the drop which baffles me.

But now, when I log in as my regular user (Not user@kasm.local) I see a really pretty background, but no apps to launch?

And when I go into the Admin account, and look at the workspaces, nothing shows up at all?

**Comment by thelamer** [comment]
@Pontiac76 It looks like for the install/upgrade logic during arch detection we use `uname -m` which should return `x86_64` or `aarch64` on a standard Linux host. It looks like your distro is returning `unknown` this will break many parts of our logic and specifically image ingestion. This explains the image ingestion problem. 

As far as upgrade output I understand that can be a bit confusing but part of the upgrade process is to run a stock install basically then restore the database backup from 1.11.0 with all of your settings along with setting the vars in these config files for future upgrades. You can see more in the manual process here: 

https://www.kasmweb.com/docs/latest/upgrade/single_server_upgrade.html#manual-upgrade

As for port assignment there is a `-L|--proxy-port` option for upgrade.sh which is required if not running on 443.

**Comment by Pontiac76** [comment]
```
stephen@Kasm:~$ uname -m
x86_64
stephen@Kasm:~$ uname
Linux
stephen@Kasm:~$ uname -a
Linux Kasm 5.10.0-19-amd64 #1 SMP Debian 5.10.149-2 (2022-10-21) x86_64 GNU/Linux
```

Seems that the expected response is there.

Going through the manual process to see what's what, I ran 

`/opt/kasm/1.12.0/bin/utils/db_init -s /opt/kasm/1.12.0/conf/database/seed_data/default_images_amd64.yaml`

Once I relogged, I saw the list of modules available, but it said that it was trying to download the modules.  I left the machine to do other things for a bit, then came back, and it seems to be alright now.

So WHY it broke is odd.  In my haste (due to work related stuff) I had closed the terminal window I was working in and didn't save the entire history of what was done.
